File: backend/what.js
// File: app.js

const express = require("express");
const bodyParser = require("body-parser");
const axios = require("axios");

const app = express();
app.use(bodyParser.json());

// 🔑 =========================================================================
// 🔑 IMPORTANT: Replace with your credentials.
// 🔑 Your token from the Meta App Dashboard EXPIRES EVERY 23 HOURS.
// 🔑 If messages stop sending, this is the first thing to replace.
// 🔑 =========================================================================
const VERIFY_TOKEN = "sanjay"; // This must match the one in your Meta App Dashboard
const WHATSAPP_TOKEN = "EAATCYRPgyw4BPmSpwBh3MmocakqZBWwI0khOwtZBnDDsK5bL3KXDbeQefmsESECN0mHB8nlgKFSHH7xDS3U3jv23WJvzlZBZAHwyZCYAAufBZCP8TMDwDqcRIYgoa18jp2324codgntlXa53RiDU9J3MR9ig5HSHyyAZBrxES2HSV7lb5ZBaksO5I3aX6lYOJ7etrNbs6yfeCDVgZCfYiPjZCD3M2DUqpXSOFUqivD21ukdwZDZD";
const PHONE_NUMBER_ID = "842332982291402"; // Your WhatsApp Business Phone Number ID

// 🤖 Your Python Chatbot Backend URL
const PYTHON_API_URL = "http://localhost:5000/ask";

// ✅ Webhook verification (for setup in Meta App Dashboard)
app.get("/webhook", (req, res) => {
  const mode = req.query["hub.mode"];
  const token = req.query["hub.verify_token"];
  const challenge = req.query["hub.challenge"];

  if (mode === "subscribe" && token === VERIFY_TOKEN) {
    console.log("✅ Webhook verified!");
    res.status(200).send(challenge);
  } else {
    console.error("❌ Webhook verification failed. Make sure your VERIFY_TOKEN matches the one in your Meta App Dashboard.");
    res.sendStatus(403);
  }
});

// ✅ Handle incoming messages and forward to Python backend
app.post("/webhook", async (req, res) => {
  try {
    const entry = req.body.entry?.[0];
    const changes = entry?.changes?.[0]?.value?.messages?.[0];

    // Ensure it's a valid text message from a user
    if (changes && changes.type === "text") {
      const from = changes.from; // User's WhatsApp number
      const msgBody = changes.text.body;

      console.log(`[IN] Message from ${from}: "${msgBody}"`);

      // 🚀 FORWARD THE MESSAGE TO THE PYTHON BACKEND
      let botResponse = "Sorry, I'm having a little trouble right now. Please try again in a moment.";

      try {
        const pythonResponse = await axios.post(PYTHON_API_URL, {
          question: msgBody,
          session_id: from, // Use the user's phone number as the unique session ID
        });

        if (pythonResponse.data && pythonResponse.data.answer) {
          botResponse = pythonResponse.data.answer;
        } else {
            console.warn("⚠️ Python API responded but had no 'answer' field.");
        }
      } catch (error) {
        console.error("❌ Error calling Python API:", error.response ? error.response.data : error.message);
        // The default error message will be used
      }
      
      console.log(`[OUT] Preparing to send to ${from}: "${botResponse}"`);

      // 📤 SEND THE PYTHON SERVER'S RESPONSE BACK TO THE USER VIA WHATSAPP
      try {
        await axios.post(
          `https://graph.facebook.com/v19.0/${PHONE_NUMBER_ID}/messages`,
          {
            messaging_product: "whatsapp",
            to: from,
            text: { body: botResponse },
          },
          { headers: { Authorization: `Bearer ${WHATSAPP_TOKEN}` } }
        );
        console.log(`✅ Message sent successfully to ${from}!`);

      } catch (error) {
        // 🔥 THIS IS THE CRITICAL DEBUGGING BLOCK 🔥
        // It will print the exact error from the WhatsApp API
        console.error("❌ FAILED TO SEND WHATSAPP MESSAGE ❌");
        if (error.response) {
          // The request was made and the server responded with a status code
          // that falls out of the range of 2xx
          console.error("Error Data:", JSON.stringify(error.response.data, null, 2));
          console.error("Error Status:", error.response.status);
          console.error("Error Headers:", error.response.headers);
        } else if (error.request) {
          // The request was made but no response was received
          console.error("Error Request:", error.request);
        } else {
          // Something happened in setting up the request that triggered an Error
          console.error("Error Message:", error.message);
        }
      }
    }
  } catch (error) {
    console.error("❌ An error occurred in the webhook handler:", error);
  }

  // Always send a 200 OK to WhatsApp to acknowledge receipt of the event
  res.sendStatus(200);
});

// ✅ Start the gateway server
app.listen(3000, () => {
  console.log("🚀 Node.js Gateway Server is running on port 3000");
  console.log("Make sure ngrok is running and your webhook is configured.");
  console.log("Waiting for messages from WhatsApp...");
});

File: backend/what.py
import os
import hmac
import hashlib
import json
from flask import Flask, request, abort
from dotenv import load_dotenv
import requests
from groq import Groq
from pyngrok import ngrok  # Add this import

load_dotenv()

app = Flask(__name__)
# Env vars
PHONENUMBERID = os.getenv('PHONENUMBERID')
ACCESS_TOKEN = os.getenv('ACCESS_TOKEN')
VERIFY_TOKEN = os.getenv('VERIFY_TOKEN')
GROQ_API_KEY = os.getenv('GROQ_API_KEY')
APP_SECRET = os.getenv('APP_SECRET')  # Optional
NGROK_AUTHTOKEN = os.getenv('NGROK_AUTHTOKEN')  # Add this
VERSION = 'v19.0'  # Update if needed

groq_client = Groq(api_key=GROQ_API_KEY)

def verify_signature(payload, signature):
    if not APP_SECRET:
        return True
    computed = hmac.new(
        APP_SECRET.encode('utf-8'),
        payload,
        hashlib.sha256
    ).hexdigest()
    return hmac.compare_digest(f"sha256={computed}", signature)

def get_ai_response(message_text):
    try:
        chat_completion = groq_client.chat.completions.create(
            model="llama3.1-70b-versatile",
            messages=[
                {"role": "system", "content": "You are a helpful WhatsApp assistant. Keep responses concise and friendly."},
                {"role": "user", "content": message_text}
            ],
            max_tokens=150,
            temperature=0.7
        )
        return chat_completion.choices[0].message.content.strip()
    except Exception as e:
        return f"Sorry, I encountered an error: {str(e)}"

def send_message(to_phone, message_text):
    url = f"https://graph.facebook.com/{VERSION}/{PHONENUMBERID}/messages"
    headers = {
        "Authorization": f"Bearer {ACCESS_TOKEN}",
        "Content-Type": "application/json"
    }
    data = {
        "messaging_product": "whatsapp",
        "to": to_phone,
        "type": "text",
        "text": {"body": message_text}
    }
    response = requests.post(url, headers=headers, json=data)
    response.raise_for_status()
    return response.json()

@app.route('/webhook', methods=['GET'])
def verify():
    mode = request.args.get('hub.mode')
    token = request.args.get('hub.verify_token')
    challenge = request.args.get('hub.challenge')
    if mode == 'subscribe' and token == VERIFY_TOKEN:
        return challenge, 200
    else:
        abort(403)

@app.route('/webhook', methods=['POST'])
def webhook():
    payload = request.get_data(as_text=True)
    signature = request.headers.get('X-Hub-Signature-256')

    if not verify_signature(payload, signature):
        abort(401)

    try:
        data = json.loads(payload)
        if 'entry' in data and len(data['entry']) > 0:
            change = data['entry'][0]['changes'][0]
            if change['field'] == 'messages' and 'value' in change:
                message = change['value']['messages'][0]
                if message['type'] == 'text':
                    from_phone = message['from']
                    user_message = message['text']['body']
                    ai_response = get_ai_response(user_message)
                    send_message(from_phone, ai_response)
        return 'OK', 200
    except Exception as e:
        print(f"Error: {e}")
        abort(400)

if __name__ == '__main__':
    # Auto-start ngrok tunnel
    if NGROK_AUTHTOKEN:
        ngrok.set_auth_token(NGROK_AUTHTOKEN)
        public_url = ngrok.connect(5000)
        print(f"ngrok tunnel started: {public_url}")
        print(f"Use this for Meta webhook: {public_url}/webhook")
    else:
        print("Warning: NGROK_AUTHTOKEN not set—run without tunnel (local only).")

    app.run(debug=True, port=5000)